### PR TITLE
Check for Missing Spaces

### PR DESCRIPTION
**Pull request by pantor**
This PR adds a `no-spaces`check for missing spaces around parameters, initializer lists and binary operators, fixing issue #4.

### CODE REVIEW

**Comment by tintou** [comment]
My $0.2, isn't it a little bit weird to call it `no-spaces` when it actually fails on missing ones?

**Comment by pantor** [comment]
I took the name from the pylint _no-space-check_, and I think it's in line with the naming of our other checks (like the DoubleSpacesCheck fails on double spaces, NoSpaceCheck fails on no or a missing space).

**Comment by pantor** [comment]
And my bad, the check is called `no-space` without `s`.

**Comment by jeremypw** [comment]
@pantor :  Conflicts needs resolving after latest commit to master.

File: lib/Check.vala [hunk 1]
```diff
@@ -89,4 +89,20 @@ public abstract class ValaLint.Check : Object {
...
+    /**
+     * Adds a mistake to the mistake list and checks for duplicates.
+     *
+     * @param mistake The mistake.
+     * @param mistakes The mistakes list.
+     */
+    protected void add_mistake (FormatMistake mistake, ref Vala.ArrayList<FormatMistake?> mistakes) {
+        mistakes.equal_func = (a, b) => {
+            return a == b;
+        };
```
Comment (jeremypw) [inline lib/Check.vala]: This seems an odd place to add an equal_func - should this not have been set when the list was created?

Comment (pantor) [reply lib/Check.vala]: This should be fixed with the latest commits.

File: lib/Checks/NoSpaceCheck.vala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                var end = expr.source_reference.end;
+
+                if (expr is Vala.Parameter) {
+                    var parameter = expr as Vala.Parameter;
+                    if (parameter.initializer != null) {
+                        end = parameter.initializer.source_reference.end;
+                    }
+                }
+
+                char char_after = * (end.pos + 1);
```
Comment (jeremypw) [inline lib/Checks/NoSpaceCheck.vala]: I do not think there should there be a space after `*` when used to dereference a pointer?

File: lib/Checks/NoSpaceCheck.vala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                    add_mistake ({ this, loc, "Missing whitespace" }, ref mistake_list);
+                }
+            }
+        }
+    }
+
+    public void check_binary_expression (Vala.BinaryExpression expr,
+                                         ref Vala.ArrayList<FormatMistake?> mistake_list) {
+
+        char char_before_operator = * (expr.left.source_reference.end.pos);
```
Comment (jeremypw) [inline lib/Checks/NoSpaceCheck.vala]: Superfluous space after `*`

File: lib/Checks/NoSpaceCheck.vala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        if (char_before_operator != ' ' && char_before_operator != '\n' && char_before_operator != ')') {
+            var loc = expr.left.source_reference.end;
+            loc.column += 1;
+            add_mistake ({ this, loc, "Missing whitespace" }, ref mistake_list);
+        }
+
+        char char_after_operator = * (expr.right.source_reference.begin.pos - 1);
+        if (char_after_operator != ' ' && char_after_operator != '\n' && char_after_operator != '(') {
```
Comment (jeremypw) [inline lib/Checks/NoSpaceCheck.vala]: Check for ',' as well?

File: lib/Checks/NoSpaceCheck.vala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+                if (expr is Vala.Parameter) {
+                    var parameter = expr as Vala.Parameter;
+                    if (parameter.initializer != null) {
+                        end = parameter.initializer.source_reference.end;
+                    }
+                }
+
+                char char_after = * (end.pos + 1);
+                if (char_after != ' ' && char_after != '\n' && char_after != ')') {
```
Comment (jeremypw) [inline lib/Checks/NoSpaceCheck.vala]: False positive given for expressions like:

`function_name ((<expression>), parameter);`

**Comment by jeremypw** [comment]
Unit test needed.

**Pull request merged**